Add "compact" elliptic DG operator

nilsvu opened the pull request:
## Proposed changes

This elliptic DG operator will replace the scheme that is currently shared between elliptic and evolution executables. By decoupling the elliptic from the evolution scheme the new DG operator doesn't have to treat the first-order "auxiliary" variables on the same footing as the "primal" variables anymore. Instead, it can directly apply to the primal variables now, essentially computing their second derivates, modified by the fluxes and sources of the elliptic system and by DG boundary corrections. This significantly reduces the size of the operator and therefore its memory usage and computational cost for solving it. Note that this new "compact" DG operator is actually equivalent to the first-order operator it replaces, in the sense that their solutions are exactly the same. It's the same scheme, but reduced to solving a smaller operator. This is facilitated by taking a Schur complement of the first-order operator, which is possible because the equations can be solved for the auxiliary variables just by inverting the DG mass matrix.

This PR only adds the DG operator and tests. A follow-up PR will switch the executables over to actually use the new DG operator, at which point the elliptic solver will be independent of the "boundary schemes" code that it currently shares with the evolution DG schemes.

I split the code in a few commits (essentially one commit per added file) to make it easier to review.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add compact elliptic DG operator

This elliptic DG operator will replace the scheme that is currently shared
between elliptic and evolution executables. By decoupling the elliptic from
the evolution scheme the new DG operator doesn't have to treat the
first-order "auxiliary" variables on the same footing as the "primal"
variables anymore. Instead, it can directly apply to the "primal" variables now,
essentially computing their second derivates, modified by the fluxes and sources
of the elliptic system and by DG boundary corrections. This significantly
reduces the size of the operator and therefore its memory usage and computational
cost for solving it. Note that this new "compact" DG operator is actually
equivalent to the first-order operator it replaces, in the sense that their
solutions are exactly the same. It's the same scheme, but reduced to solving
a smaller operator. This is facilitated by taking a Schur complement of the
first-order operator, which is possible because the auxiliary equation can
be solved for the auxiliary variables just by inverting the DG mass matrix.

nilsvu committed:
Add actions for elliptic DG operator

nilsvu committed:
Add tests for elliptic DG operator

nilsvu commented:
I moved the first two commits to #2913 to make this PR slightly smaller :)

nilsvu commented:
@sxs-collaboration/spectre-core-devs Since this PR is relatively involved I don't want to "request" reviews from anyone, but I really need this moving forward to get the elliptic solver merged. Would two of you be willing to take this on?

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -0,0 +1,945 @@
+ * u^\mathrm{ext}\right) \\
+ * (n_i F^i_{v_j})^* &= n_j u^* = \frac{1}{2} n_j \left(u^\mathrm{int} +
+ * u^\mathrm{ext}\right)
+ * \f}
```
This pair of equations is in the opposite order from the corresponding pair above.  If there's no reason for that, swap one pair so they match.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -0,0 +1,945 @@
+  return ::dg::needs_projection(face_mesh, mortar_mesh, mortar_size)
+             ? boundary_data.project_to_mortar(face_mesh, mortar_mesh,
+                                               mortar_size)
+             : boundary_data;
```
Does this need a `std::move`?  I'm not sure.

nilsvu replied:
Yes I think so. If I understand correctly, copy elision only works when returning a named or temporary object:

> Standard, 12.8.32:
> - in a return statement in a function with a class return type, when the expression is the name of a non-volatile automatic object (other than a function or catch-clause parameter) with the same cv-unqualified type as the function return type, the copy/move operation can be omitted by constructing the automatic object directly into the function's return value

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -0,0 +1,945 @@
+              tmpl::list<Tags::NormalDotFluxForJump<PrimalMortarVars>...>>();
+      primal_boundary_corrections_on_mortar -=
+          remote_data.field_data.template extract_subset<
+              tmpl::list<Tags::NormalDotFluxForJump<PrimalMortarVars>...>>();
```
Every `extract_subset` allocates.  I don't know if being able to do math on the resulting `Variables` outweighs that or not, but it's probably a good opportunity for another "possible optimization" comment like above.

nilsvu replied:
Probably not, I pushed a fixup that changes these to parameter-pack expansions and loops over components. Please let me know if you think this is a good change or if I should rather drop the fixup, I'm undecided.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+            const auto... fields_and_fluxes) noexcept {
+          ASSERT(boundary_conditions.contains(direction),
+                 "No boundary condition is available in block " +
+                     std::to_string(element_id.block_id()) + " in direction " +
```
You can use  `<<` in ASSERT messages.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+    // `db::mutate`, so we retrieve the pointers to the memory buffers in
+    // advance. This should be safe since the DataBox is not manipulated while
+    // the pointers are being used (no `db::create_from`, serialization etc.),
+    // but we have to take care of propagating mutations to the data to subtimes
```
"subitems"

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+    // the pointers are being used (no `db::create_from`, serialization etc.),
+    // but we have to take care of propagating mutations to the data to subtimes
+    // (see below).
+    const auto [primal_fluxes, all_mortar_data] =
+        db::mutate<PrimalFluxesTag, all_mortar_data_tag>(
+            make_not_null(&box),
+            [](const auto... args) { return std::make_tuple(args...); });
```
This use was declared forbidden when `db::mutate` return values were added.  See the documentation.

wthrowe replied:
Maybe move the objects out of the box, mutate them, and then move them back in?

nilsvu replied:
Added a fixup that does that. Do you think that's preferable to retrieving the pointers?

wthrowe replied:
In a case where there might be compute items involved, I think it is better to always keep the box in a consistent state by moving the data out.  For the specific case where there are known to be no compute items I'm less sure, but there was a consensus among the reviewers on #2933 that pulling out pointers was bad, so I'll follow that until there's a reason not to.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+
+    // Can't `db::get` the arguments for the boundary conditions within
+    // `db::mutate`, so we retrieve the pointers to the memory buffers in
+    // advance. This should be safe since the DataBox is not manipulated while
```
Could you pass the const stuff into the mutate, like is usually done?  It doesn't look like you need the box itself in your mutation, which is usually the only case where this sort of workaround is needed.

nilsvu replied:
I only need the box in the mutation to retrieve arguments for the boundary conditions, in the `apply_boundary_condition` lambda. The boundary conditions are selected at runtime, so the `elliptic::apply_boundary_condition` function uses `call_with_dynamic_type` to loop over the derived classes, choose the right one and call it with arguments retrieved from the box. The only alternative to passing the box around that I can think of right now is retrieving the arguments in advance in another `call_with_dynamic_type`, storing const-refs in a large `std::variant` (one variant for every registered boundary-condition class) and passing that to `prepare_mortar_data`.

wthrowe replied:
OK, yep, I missed the explicit use of the box.  Fine as-is.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+        make_not_null(&box),
+        [](const auto operator_applied_to_fields, const auto all_mortar_data,
+           const auto& primal_fields, const auto& primal_fluxes,
+           const auto&... args) noexcept {
```
Why explicitly list and forward some arguments, but not all of them?  (Or, can `elliptic::dg::apply_operator<System, Linearized>` be used as the argument to `db::mutate` directly?)

nilsvu replied:
Pushed a fixup that forwards all const-ref args. Is there a way to get rid of the lambda entirely, or forward all arguments including the not-null pointers? Just passing `&elliptic::dg::apply_operator<System, Linearized>` instead of the lambda doesn't work because the function infers its template parameters from arguments.

wthrowe replied:
Sometimes you can get rid of the lambda with a cast to a function pointer type, but I don't see a way to work out what the correct type to cast to is here.  You can forward the pointers by just removing the explicit arguments and absorbing them into the argument pack.

nilsvu replied:
Of course 🤦‍♂️ the pointers can also just be forwarded as const-refs..

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+
+/*!
+ * \brief Apply the DG operator to the `PrimalFieldsTag` and write the result to
+ * the `OperatorAppliedToFieldsTag`
```
I can't find this documentation.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+ * You can specify the `PrimalMortarFieldsTag` and the `PrimalMortarFluxesTag`
+ * to re-use mortar-data memory buffers from other operator applications, for
+ * example when applying the nonlinear and linearized operator. They default to
+ * the `PrimalFieldsTag` and the `PrimalFluxesTag`.
```
What is the downside to using the default?  Does it overwrite the primal fluxes mentioned in the previous paragraph?

nilsvu replied:
The linear and nonlinear operator apply to different variables, so when using the defaults here they each set up memory buffers for their mortar data. However, when the operator applications don't overlap in the algorithm, as is the case for the Newton-Krylov algorithm, we can just re-use the same buffers in the DataBox for both the linear and nonlinear operator. In that case we can set these template parameters to the same tags. I extended the docs here a little bit.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:
```diff
@@ -0,0 +1,629 @@
+/// \cond
+template <typename System, typename FixedSourcesTag, typename... FluxesArgsTags,
+          typename... SourcesArgsTags>
+struct ImposeInhomogeneousBoundaryConditionsOnSource<
```
Pretty much the same comments as for PrepareAndSendMortarData above.

wthrowe reviewed (changes requested):
I've done this first two commits, haven't done the third yet.

wthrowe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -0,0 +1,668 @@
+  if constexpr (Linearized) {
+    INFO(
+        "Test imposing inhomogeneous boundary conditions on source, i.e. b -= "
+        "A(x=0)");
```
Are there any checks in this block?  If not, an `INFO` doesn't do anything.

nilsvu replied:
The `INFO` also prints when Catch2 fails a test for other reasons, such as uncaught exceptions. I think I put this in because I had a bug here at some point, but I don't remember exactly. I can take it out if you prefer.

wthrowe replied:
OK, that's fine.

wthrowe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -0,0 +1,668 @@
+        Poisson::FirstOrderSystem<2, Poisson::Geometry::FlatCartesian>;
+    Poisson::Solutions::ProductOfSinusoids<2> analytic_solution{{{M_PI, M_PI}}};
+    Approx analytic_solution_approx =
+        Approx::custom().epsilon(1.e-2).margin(12.);
```
The allowed errors for these tests are huge.  This margin is an order of magnitude larger than many of the values to be checked.

nilsvu replied:
I'll do a few more tests to look at the error margins here.

wthrowe reviewed (changes requested):
Review of third commit.

nilsvu commented:
I appreciate you finding the time to review this @wthrowe! I pushed fixups to address most of your comments and commented on the others.

nilsvu commented:
Pushed two more fixup. Thanks @wthrowe!

wthrowe commented:
I'm happy with the current state except for the open question about test tolerances.  Go ahead and squash the rest.

nilsvu commented:
Rebased and squashed the fixups. I pushed another fixup that addresses the issue with test tolerances. It adjusts the scales and adds additional higher-resolution analytic-solution tests with stricter tolerances. Please take a look when you get the chance :)

wthrowe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -383,256 +387,341 @@ SPECTRE_TEST_CASE("Unit.Elliptic.DG.Operator", "[Unit][Elliptic]") {
+          Approx::custom().epsilon(8.e-1).scale(M_PI);
+      Approx analytic_solution_operator_approx =
+          Approx::custom().epsilon(8.e-1).scale(M_PI * penalty_parameter *
+                                                square(4) / 2.);
```
I'm still worried that you need to accept 80% errors in this case (and slightly less above).  Why is this result so inaccurate?

nilsvu replied:
Well this 3D case tries to resolve a full sinusoid period with 2 linear elements in xi-direction :) At x=-0.5 the gradient is zero analytically, but 2 numerically because the sinusoid is approximated as a ^ shape (or upside-down V, i.e. just two lines with slope 2 and -2). That's a large error. I chose so few elements to avoid writing even more numbers in the regression test. The added code below repeats the analytic-solution test with much higher resolution. Note that the low tolerance here only applies to the analytic-solution test; the custom values are checked to default precision (1.-14 or so).

wthrowe replied:
Can you set the expected values to what you actually expect the operator to return, then?

nilsvu replied:
Yea that's what the regression test does on the low-resolution domain. I could disable the analytic-solution test on the low-resolution domain by making the `approx` arguments `std::optional`s, so the comparison to the analytic solution only runs on the high-resolution domain. Not sure if that's worth it though, it just deletes these low-resolution/high-tolerance analytic-solution tests.

wthrowe replied:
Sorry, you've lost me.  This *is* the regression test, isn't it?

nilsvu replied:
Yes it is. It compares the hard-coded numbers to operator output at default precision (1e-14 or so). It is at low resolution to keep the amount of hard-coded numbers low, and because the resolution is irrelevant for the regression test. It also does an analytic-solution test because that's built-in to the testing function, but does it at very low resolution and thus very high tolerance. Therefore, as response to your comment, I added another test at much higher resolution below. That one doesn't test any hard-coded numbers but only the analytic solution at much smaller tolerance.

wthrowe replied:
Oh, I see now.  Could you add a comment explaining that these loose approx values are effectively disabling a part of the test you don't want, but there is another part running with more typical tolerances?  I didn't understand that and thought these were the only tolerances being used.

nilsvu replied:
Done! Pushed another fixup.

wthrowe replied:
I think that's clear, now.  Go ahead and squash.

nilsvu commented:
Rebased on #3048.

nilsvu commented:
Rebased again :)

nilsvu commented:
@kidder could you perhaps do the second review for this or help find a second reviewer?

nilsdeppe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -0,0 +1,976 @@
+ * further details here, the discretization results in a linear equation
+ * \f$A(u)=b\f$ over all grid points and primal variables. Solving the elliptic
+ * equations amounts to numerically inverting the DG operator \f$A\f$, typically
+ * without every constructing the full matrix but by employing an iterative
```
`every` -> `ever`

nilsdeppe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -0,0 +1,976 @@
+      // (as opposed to the LLF flux) uses the raw field derivatives without
+      // boundary corrections in the average, which is why we can communicate
+      // the data so early together with the auxiliary boundary data. In this
+      // case the penalty needs to include a factor p^2 / h (see the `penalty`
```
the documentation said you use `N_points`, is `p=N_points` here? That would be good to clarify :)

nilsdeppe reviewed (dismissed):
LGTM, great documentation! A couple minor comments, feel free to ignore :)

nilsvu commented:
Rebased and fixed those two things. This is a fairly involved PR, so your reviews are much appreciated @wthrowe and @nilsdeppe :)
